pkg/core/interop_neo.go

```diff
@@ -317,30 +288,46 @@ func contractIsPayable(ic *interop.Context, v *vm.VM) error {
 
 // contractMigrate migrates a contract.
 func contractMigrate(ic *interop.Context, v *vm.VM) error {
+	currHash := v.GetContextScriptHash(0) // current script hash
+	contract, err := ic.DAO.GetContractState(currHash)
+	if contract == nil {
+		return errors.New("contract doesn't exist")
+	}
 	newcontract, err := createContractStateFromVM(ic, v)
 	if err != nil {
 		return err
 	}
-	contract, err := ic.DAO.GetContractState(newcontract.ScriptHash())
-	if err != nil {
-		contract = newcontract
-		err := ic.DAO.PutContractState(contract)
+	if newcontract.Script != nil {
+		if l := len(newcontract.Script); l == 0 || l > MaxContractScriptSize {
+			return errors.New("invalid script len")
+		}
+		h := newcontract.ScriptHash()
+		if h.Equals(contract.ScriptHash()) {
+			return errors.New("the script is the same")
+		} else if _, err := ic.DAO.GetContractState(h); err == nil {
+			return errors.New("contract already exists")
+		}
+		newcontract.ID = contract.ID
+		if err := ic.DAO.PutContractState(newcontract); err != nil {
+			return err
+		}
+		if err := ic.DAO.DeleteContractState(contract.ScriptHash()); err != nil {
+			return err
+		}
+	}
+	if contract.HasStorage() {
+		// TODO store items by ID
```

Comment:
Issue #?

Reply:
It actually is an important one as it directly affects state calculations.

Reply:
https://github.com/nspcc-dev/neo-go/issues/1037